### PR TITLE
Display banner when viewing/reviewing apps with no/incorrect competency

### PR DESCRIPTION
**Pull request by hiboabd**
## Description of change

Caseworkers who do not have competencies/ do not have the required competency to work on an application are now prevented from doing so by displaying a notification banner. This impacts the following actions: 

When a caseworker attempts to: 
	- Review an application (assign/reassign) 
	- View the open/closed applications page
	- Retrieve an application (get next)

## Link to relevant ticket
[CRIMAPP-279](https://dsdmoj.atlassian.net/browse/CRIMAPP-279)

### After changes:
**As a caseworker with no competencies set** 

<table>
  <tr>
    <th>Test case</th>
    <th>After change</th>
  </tr>
  <tr>
    <td>When attempting to reassign an application to self</td>
    <td>Screenshot 2023-12-14 at 17 41 38</td>
  </tr>
  <tr>
    <td>When attempting to assign an application to self</td>
    <td>Screenshot 2023-12-14 at 17 41 16</td>
  </tr>
  <tr>
    <td>When attempting to retrieve an application</td>
    <td>Screenshot 2023-12-14 at 17 41 00</td>
  </tr>
  <tr>
    <td>When viewing open applications tab</td>
    <td>Screenshot 2023-12-14 at 17 40 48</td>
  </tr>
  <tr>
    <td>When viewing closed applications tab</td>
    <td>Screenshot 2023-12-14 at 17 40 53</td>
  </tr>
</table>

**As a caseworker with a competency that is different from the application being viewed**

<table>
  <tr>
    <th>Test case</th>
    <th>After change</th>
  </tr>
  <tr>
    <td>When attempting to reassign an application to self</td>
    <td>Screenshot 2023-12-14 at 17 45 37</td>
  </tr>
  <tr>
    <td>When attempting to assign an application to self</td>
    <td>Screenshot 2023-12-14 at 18 14 14</td>
  </tr>
</table>

## How to manually test the feature

Test the following actions when logged in as a caseworker with no competencies set 
	- Assigning/reassigning an application to self (will need to use the search to find an application)
	- Viewing the open/closed applications pages
	- Retrieving an application (get next)

Test the following actions when logged in as a caseworker with a competency that is different from the application being viewed
	- 	Assigning/reassigning an application to self (will need to use the search to find an application)

[CRIMAPP-279]: https://dsdmoj.atlassian.net/browse/CRIMAPP-279?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: config/locales/en/casework.yml [hunk 1]
```diff
@@ -38,6 +38,16 @@ en:
       cannot_send_back_when_marked_as_ready: This application was already marked as ready for assessment
       cannot_download_doc_uploaded_to_another_app: File must be uploaded to current application to download
       cannot_download_try_again: "%{file_name} could not be downloaded – try again"
+      cannot_view_apps_without_a_work_stream:
+        - You must be allocated to a work queue to view applications
```
Comment (hiboabd) [inline config/locales/en/casework.yml:42]: Probably need to change work queue but type of work sounds awkward here

File: app/controllers/casework/reassigns_controller.rb [hunk 2]
```diff
@@ -1,20 +1,21 @@
 module Casework
   class ReassignsController < Casework::BaseController
+    before_action :set_crime_application
     def new
       @current_assignment = current_assignment
     end
 
-    def create
+    def create # rubocop:disable Metrics/AbcSize,
+      return require_work_stream_flash if current_user.work_streams.empty?
```
Comment (willmcb) [inline app/controllers/casework/reassigns_controller.rb:9]: Think these lines are identical to https://github.com/ministryofjustice/laa-review-criminal-legal-aid/pull/486/files#diff-25055c48cc370e7cd6c7a09667ccddde6f433f21207db50570d7d65bca459c1eR16

Perhaps we could move them to a concern and then call them in a before action for each?

File: app/controllers/casework/reassigns_controller.rb [hunk 3]
```diff
@@ -35,5 +36,14 @@ def current_assignment
         assignment_id: params[:crime_application_id]
       )
     end
+
+    def require_work_stream_flash
```
Comment (willmcb) [inline app/controllers/casework/reassigns_controller.rb:40]: can we make these generic? if we specified constants in the controllers themselves? Just a thought :)

File: config/locales/en/casework.yml (hunk 1 above)
Comment (willmcb) [reply config/locales/en/casework.yml:42]: Agree

File: spec/system/casework/assigning/assign_an_application_to_myself_spec.rb [hunk 4]
```diff
@@ -3,6 +3,9 @@
 RSpec.describe 'Assigning an application to myself' do
   include_context 'with an existing application'
   let(:assign_cta) { 'Assign to your list' }
+  let(:banner_text) do
+    "You must be allocated to the Cat 1 work queue to review this application\nContact your supervisor to arrange this"
```
Comment (willmcb) [inline spec/system/casework/assigning/assign_an_application_to_myself_spec.rb:7]: Is it "Cat 1" or "CAT1"? If its an acronynm we should keep it as caps I think.

File: spec/system/casework/assigning/reassign_an_application_to_myself_spec.rb [hunk 5]
```diff
@@ -1,28 +1,13 @@
...
 
-  let(:assignee) do
-    User.create(
-      first_name: 'Fred',
-      last_name: 'Smitheg',
-      auth_oid: SecureRandom.uuid,
-      email: 'Fred.Smitheg@justice.gov.uk'
-    )
+  let(:banner_text) do
+    "You must be allocated to the Cat 1 work queue to review this application\nContact your supervisor to arrange this"
```
Comment (willmcb) [inline spec/system/casework/assigning/reassign_an_application_to_myself_spec.rb:7]: Cat1 v CAT1

**Review by timpeat — APPROVED** [review]
Great start. I'd be happy for this to go in as is for user testing.

File: app/controllers/casework/reassigns_controller.rb [hunk 6]
```diff
@@ -35,5 +36,14 @@ def current_assignment
...
+      set_flash(:no_work_streams_to_assign_from, success: false)
+    end
+
+    def require_app_work_stream_flash
+      set_flash(:not_allocated_to_appropriate_work_stream,
+                success: false, work_queue: WorkStream.new(@crime_application.work_stream).to_param.humanize)
```
[the remark this answers is not in the document]
Comment (timpeat) [reply app/controllers/casework/reassigns_controller.rb:46]: I've started, but not finished a bit of a refactor here: https://github.com/ministryofjustice/laa-review-criminal-legal-aid/tree/CRIMAPP-279-no-competencies-flow-tim

File: spec/system/casework/assigning/reassign_an_application_to_myself_spec.rb (hunk 5 above)
Comment (hiboabd) [reply spec/system/casework/assigning/reassign_an_application_to_myself_spec.rb:7]: I prefer CAT 1

File: spec/system/casework/assigning/assign_an_application_to_myself_spec.rb (hunk 4 above)
Comment (hiboabd) [reply spec/system/casework/assigning/assign_an_application_to_myself_spec.rb:7]: It's a acronym so will change to caps

File: app/controllers/casework/reassigns_controller.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (hiboabd) [reply app/controllers/casework/reassigns_controller.rb:46]: Thanks tim :)

**Pull request merged**